sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/BeamFnDataBufferingOutboundObserver.java

```diff
@@ -0,0 +1,129 @@
+
+/**
+ * A buffering outbound {@link Consumer} for the Beam Fn Data API.
+ *
+ * <p>Encodes individually consumed elements with the provided {@link Coder} producing
+ * a single {@link org.apache.beam.fn.v1.BeamFnApi.Elements} message when the buffer threshold
+ * is surpassed.
+ *
+ * <p>The default buffer threshold can be overridden by specifying the experiment
+ * {@code beam_fn_api_data_buffer_limit=<bytes>}
+ */
+public class BeamFnDataBufferingOutboundObserver<T>
+    implements CloseableThrowingConsumer<WindowedValue<T>> {
+  private static final String BEAM_FN_API_DATA_BUFFER_LIMIT = "beam_fn_api_data_buffer_limit=";
+  private static final int DEFAULT_BUFFER_LIMIT = 1_000_000;
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(BeamFnDataBufferingOutboundObserver.class);
+
+  private long byteCounter;
+  private long counter;
+  private final int bufferLimit;
+  private final Coder<WindowedValue<T>> coder;
+  private final KV<Long, BeamFnApi.Target> outputLocation;
+  private final StreamObserver<BeamFnApi.Elements> outboundObserver;
+  private final ByteString.Output bufferedElements;
+
+  public BeamFnDataBufferingOutboundObserver(
+      PipelineOptions options,
+      KV<Long, BeamFnApi.Target> outputLocation,
+      Coder<WindowedValue<T>> coder,
+      StreamObserver<BeamFnApi.Elements> outboundObserver) {
+    this.bufferLimit = getBufferLimit(options);
+    this.outputLocation = outputLocation;
+    this.coder = coder;
+    this.outboundObserver = outboundObserver;
+    this.bufferedElements = ByteString.newOutput();
+  }
+
+  /**
+   * Returns the {@code beam_fn_api_data_buffer_limit=<int>} experiment value if set. Otherwise
+   * returns the default buffer limit.
+   */
+  private static int getBufferLimit(PipelineOptions options) {
+    List<String> experiments = options.as(DataflowPipelineDebugOptions.class).getExperiments();
+    for (String experiment : experiments == null ? Collections.<String>emptyList() : experiments) {
+      if (experiment.startsWith(BEAM_FN_API_DATA_BUFFER_LIMIT)) {
+        return Integer.parseInt(experiment.substring(BEAM_FN_API_DATA_BUFFER_LIMIT.length()));
+      }
+    }
+    return DEFAULT_BUFFER_LIMIT;
+  }
+
+  @Override
+  public void close() throws Exception {
+    BeamFnApi.Elements.Builder elements = convertBufferForTransmission();
+    // This will add an empty data block representing the end of stream.
+    elements.addDataBuilder()
+        .setInstructionReference(outputLocation.getKey())
+        .setTarget(outputLocation.getValue());
+
+    LOGGER.debug("Closing stream for instruction {} and "
+        + "target {} having transmitted {} values {} bytes",
+        outputLocation.getKey(),
+        outputLocation.getValue(),
+        counter,
+        byteCounter);
+    outboundObserver.onNext(elements.build());
+  }
+
+  @Override
+  public void accept(WindowedValue<T> t) throws IOException {
+    coder.encode(t, bufferedElements, Context.NESTED);
```

Comment:
this would have a problem with elements that encode to 0 bytes even in the nested context, right? This is why `WindowedValueCoder` is never allowed to do so?